web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
-		}
-		var teacher int
-		if strings.HasSuffix(provider, TeacherSuffix) {
-			teacher = 1
+			if err := sessionStore.Save(r, w, s); err != nil {
+				authenticationError(logger, w, fmt.Sprintf("failed to create new session (%s): %s", SessionKey, err))
+				return
+			}
 		}
-		logger.Debugf("Provider: %v ; Teacher: %v", provider, teacher)
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect := extractRedirectURL(r, Redirect)
-		logger.Debugf("redirect: %v", redirect)
-		// TODO: Add a random string to protect against CSRF.
-		qv.Set(State, strconv.Itoa(teacher)+redirect)
-		logger.Debugf("State: %v", strconv.Itoa(teacher)+redirect)
-		r.URL.RawQuery = qv.Encode()
-		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
+		// Get provider name.
+		provider := GetProviderName(r.URL.Path, 2)
+		if provider == "" {
+			authenticationError(logger, w, "incorrect request URL")
+			return
+		}
+		providerConfig, ok := authConfig.providers[provider]
+		if !ok {
+			authenticationError(logger, w, "provider is not enabled")
+			return
+		}
 
-		url, err := gothic.GetAuthURL(w, r)
-		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		// Check teacher suffix, update scopes if teacher.
+		// Won't be necessary with GitHub App.
+		if strings.Contains(r.URL.Path, TeacherSuffix) {
+			logger.Debug("found teacher suffix on login")
+			providerConfig.Scopes = teacherScopes
+		} else {
+			providerConfig.Scopes = studentScopes
 		}
-		logger.Debugf("Redirecting to %s to perform authentication; AuthURL: %v", provider, url)
-		return c.Redirect(http.StatusTemporaryRedirect, url)
+		logger.Debugf("Provider callback URL: %s", providerConfig.RedirectURL)
+		redirectURL := providerConfig.AuthCodeURL(secret)
+		logger.Debugf("redirecting to AuthURL: %v", redirectURL)
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
 	}
 }
 
 // OAuth2Callback handles the callback from an oauth2 provider.
-func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, scms *Scms) echo.HandlerFunc {
-	return func(c echo.Context) error {
+func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, authConfig *AuthConfig, scms *Scms, secret string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
 		logger.Debug("OAuth2Callback: started")
-		w := c.Response()
-		r := c.Request()
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect, teacher := extractState(r, State)
-		logger.Debugf("Redirect: %v ; Teacher: %t", redirect, teacher)
+		if r.Method != "GET" {
+			authenticationError(logger, w, fmt.Sprintf("illegal request method: %s", r.Method))
+			return
+		}
 
-		provider, err := gothic.GetProviderName(r)
-		if err != nil {
-			logger.Error("failed to get gothic provider", zap.Error(err))
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		// Get provider.
+		provider := GetProviderName(r.URL.Path, 3)
```

Comment:
Why 3 here, when 2 was used in `OAuthLogin()`

Reply:
What's problem with?:
```go
provider, err := gothic.GetProviderName(r)
```

Reply:
Endpoint for `OAuth2Login` is `/auth/<provider>`. Endpoint for `OAuth2Callback` is `/auth/callback/<provider>`.

All `gothic` interactions rely heavily on custom provider data saved in the user session. Should I try to fine-tune sessions here if I am planning on replacing them with JWTs in the following PR?

Reply:
Should we change `/auth/callback/<provider>` to `/callback/provider`?

Reply:
I think we don't need the `GetProviderName()` approach at all, if we just use the approach outlined in [my comment](https://github.com/quickfeed/quickfeed/pull/692/files#r921802570).

Reply:
The point is that we don't want to support more than one provider per running instance of quickfeed. If you want to run with gitlab, then you need to deploy another instance of quickfeed (with a separate database etc.)

The old approach of supporting multiple providers was a design decision that we should have dropped long ago.